kubelet: DRA: fix cache integrity

bart0sh opened the pull request:
#### What type of PR is this?

/kind bug
/kind cleanup

#### What this PR does / why we need it:

Changed ClaimInfo and claimInfoCache APIs to avoid updating cached claim infos when the cache is not locked.
This should fix possible data race conditions caused by changing cached claim infos.

Added new unit and e2e_node test cases to test cache integrity.

#### Which issue(s) this PR fixes:

Fixes #123474 #116411

#### Special notes for your reviewer:

This is a second attempt to fix #123474. See [previous PR](https://github.com/kubernetes/kubernetes/pull/124284) and  [this review comment](https://github.com/kubernetes/kubernetes/pull/124284#discussion_r1565406305) for more details.

#### Does this PR introduce a user-facing change?
```release-note
NONE
```

bart0sh commented:
/sig node

bart0sh commented:
/cc @pohly @klueska

pohly commented on pkg/kubelet/cm/dra/claiminfo.go:
```diff
@@ -31,32 +31,27 @@ import (
+func (info *ClaimInfo) deepCopy() ClaimInfo {
+	claimInfo := ClaimInfo{
+		ClaimInfoState: info.ClaimInfoState.DeepCopy(),
+		annotations:    make(map[string][]kubecontainer.Annotation),
```
You can use `maps.Clone` here and then don't need the loop.

pohly commented on pkg/kubelet/cm/dra/manager.go:
```diff
@@ -123,15 +125,17 @@ func (m *ManagerImpl) PrepareResources(pod *v1.Pod) error {
 		// started, so it's OK (actually correct) to not include it
 		// in the cache.
-		claimInfo.addPodReference(pod.UID)
+		if err := m.cache.addPodReference(*claimName, pod.Namespace, pod.UID); err != nil {
```
Doesn't this still have the same problem that I [outlined before](https://github.com/kubernetes/kubernetes/pull/124284#discussion_r1565414670) (pod gets added to cache and potentially flushed to disk *before* the NodePrepareResources actually succeeds)?

bart0sh replied:
Thank you for the review. done.

bart0sh commented on pkg/kubelet/cm/dra/state/state_checkpoint.go:
```diff
@@ -63,6 +63,44 @@ type ClaimInfoState struct {
+		Namespace:       cis.Namespace,
+		PodUIDs:         cis.PodUIDs.Clone(),
+		ResourceHandles: DeepCopyResourceHandles(cis.ResourceHandles),
+		CDIDevices:      DeepCopyCDIDevices(cis.CDIDevices),
```
@pohly Is it safe to use `maps.Clone` here and remove `DeepCopyCDIDevices`?

klueska replied:
I would leave it for symmetry, but call maps.Clone inside the body of `DeepCopyCDIDevices`

pohly replied:
Agreed.

bart0sh replied:
`maps.Clone` returns a shallow copy, which may be confusing to have it inside `DeepCopyCDIDevices`, but if you insist I'll do as you've suggested.

klueska replied:
I don't think its confusing at all -- its about encapsulation -- if at some point in the future CDIDevices got more complicated than it is today (i.e. additional fields beyond just a map), you would need to also copy those fields (possibly deeply).

klueska commented on pkg/kubelet/cm/dra/claiminfo.go:
```diff
@@ -112,7 +119,7 @@ func newClaimInfo(driverName, className string, claimUID types.UID, claimName, n
 
 // newClaimInfoFromResourceClaim creates a new ClaimInfo object
-func newClaimInfoFromResourceClaim(resourceClaim *resourcev1alpha2.ResourceClaim) *ClaimInfo {
+func newClaimInfoFromResourceClaim(resourceClaim *resourcev1alpha2.ResourceClaim) ClaimInfo {
```
Why this change? Normally "new" calls return a pointer.

bart0sh replied:
This is done to make this similar to `getClaimInfo`, which returns ClaimInfo value.

klueska replied:
I would still argue to return a pointer here for consistency, but I don't feel too strongly given how it's used in this component to assign to a variable that either gets its value from this function or one that (necessarily) returns a non-pointer type.

bart0sh replied:
I don't feel too strong about it either. I just thought that this code:
```
claimInfoCopy, err := m.cache.getClaimInfo(*claimName, pod.Namespace)
if err != nil {
	// claim does not exist in cache, create new claimInfo object
	// to be processed later.
	claimInfoCopy = newClaimInfoFromResourceClaim(resourceClaim)
}
```
would look a little bit more consistent if both `get` and `newClaimInfoFromResourceClaim` functions would return a value.

bart0sh replied:
one more point: if `newClaimInfoFromResourceClaim` we should probably test it for nil or assume it never returns nil.

klueska replied:
Let's leave it as a non-pointer

klueska commented on pkg/kubelet/cm/dra/claiminfo.go:
```diff
@@ -179,11 +186,18 @@ func (cache *claimInfoCache) add(res *ClaimInfo) {
 	cache.claimInfo[res.ClaimName+res.Namespace] = res
 }
 
-func (cache *claimInfoCache) get(claimName, namespace string) *ClaimInfo {
+// getClaimInfo returns a deep copy of the cached claimInfo object
+// for the specified claim name and namespace
+func (cache *claimInfoCache) getClaimInfo(claimName, namespace string) (ClaimInfo, error) {
```
Why the name change? claimInfoCache.GetClaimInfo() seems unnecessarily redundant.

bart0sh replied:
ClaimInfoCache also contains state and usually used as m.cache, so it's not always obvious what exactly get function returns if you look at the manager code. I can change it back if you think it doesn't make sense or can rename to say getClaimInfoCopy to emphasise that it returns a clone/copy. WDYT?

klueska replied:
I don't buy the argument about containing a `state` variable. The `state` variable is internal and only used for syncing claimInfos to the checkpoint. The purpose of the abstraction is to cache claimInfo objects via get/put calls, so calling the function `get()` seems appropriate and correct.

If you think calling `m.cache.get()` is not clear, then I'd argue for a change in the `cache` variable name, not the function call name, i.e. `m.claimInfoCache.get()`

klueska commented on pkg/kubelet/cm/dra/claiminfo.go:
```diff
@@ -179,11 +186,18 @@ func (cache *claimInfoCache) add(res *ClaimInfo) {
 	defer cache.RUnlock()
 
-	return cache.claimInfo[claimName+namespace]
+	claimInfo, ok := cache.claimInfo[claimName+namespace]
+	if !ok {
+		return ClaimInfo{}, fmt.Errorf("unable to get resource for namespace: %s, claim: %s", namespace, claimName)
+	}
+
+	return claimInfo.deepCopy(), nil
```
This seems to be is the main difference from what we were doing before. Since this always returns a copy we don't have to worry about accesses to it outside of the cache.

bart0sh replied:
Yes, this and the correspondent changes in the manager.go caused by this change.

klueska commented on pkg/kubelet/cm/dra/manager.go:
```diff
@@ -108,11 +108,11 @@ func (m *ManagerImpl) PrepareResources(pod *v1.Pod) error {
 
-		claimInfo := m.cache.get(*claimName, pod.Namespace)
-		if claimInfo == nil {
+		claimInfoCopy, err := m.cache.getClaimInfo(*claimName, pod.Namespace)
+		if err != nil {
```
I don't think this rename is necessary. The new samantics of the `get` call make it clear that its a copy.

bart0sh replied:
I'd agree if we rename `get` -> `getCopy` to emphasise it even more.

klueska replied:
Regardless of the function name change, I still think this rename of the variable is unnecessary

klueska commented on pkg/kubelet/cm/dra/manager.go:
```diff
@@ -123,15 +123,16 @@ func (m *ManagerImpl) PrepareResources(pod *v1.Pod) error {
-		if claimInfo.prepared {
+		if claimInfoCopy.prepared {
 			// Already prepared this claim, no need to prepare it again
+			m.cache.add(&claimInfoCopy)
```
It feels weird to call a function named `add` here. Is the purpose of this to overwrite the existing claimInfo entry with the one that has the updated pod reference? Maybe an `update` call would make that clearer?

bart0sh replied:
I agree, but this call adds or replaces value in the map. How about addOrUpdate or something similar?

klueska replied:
`addOrUpdate()` seems reasonable

klueska commented on pkg/kubelet/cm/dra/manager.go:
```diff
@@ -149,7 +150,7 @@ func (m *ManagerImpl) PrepareResources(pod *v1.Pod) error {
 			}
 			batches[pluginName] = append(batches[pluginName], claim)
 		}
-		claimInfos[resourceClaim.UID] = claimInfo
+		claimInfos[resourceClaim.UID] = &claimInfoCopy
```
Remind me what the difference between this claimInfos slice and the claimInfo cache is. Why is this not just a call to `m.cache.add(&claimInfoCopy)` (or `update()` as I suggested above)?

bart0sh replied:
Objects in this map(new claim infos or copies of the cached ones) are not cached. They're used to to perform batched calls of the Node[Un]PrepareResources GRPC. They're added to the cache only after calls succeed. BTW, this can be a valid fix for https://github.com/kubernetes/kubernetes/issues/116411.

klueska replied:
Ack, this seems correct then. I'd need to look closer at #116411 to see how this is related

klueska commented on pkg/kubelet/cm/dra/manager.go:
```diff
@@ -277,21 +278,17 @@ func (m *ManagerImpl) GetResources(pod *v1.Pod, container *v1.Container) (*Conta
-			if claimInfo == nil {
-				return nil, fmt.Errorf("unable to get resource for namespace: %s, claim: %s", pod.Namespace, *claimName)
+			claimInfo, err := m.cache.getClaimInfo(*claimName, pod.Namespace)
+			if err != nil {
+				return nil, err
```
Why drop the extra error information?

bart0sh replied:
It's not dropped, it's moved to the `cache.getClaimInfo` since it returns error now.

klueska replied:
I'd still argue you need to wrap it here for consistency, especially since (by convention) it is the *caller's* responsibility to provide the namespace and claimname in the error string (not the *callee*).

klueska commented on pkg/kubelet/cm/dra/manager.go:
```diff
@@ -333,6 +330,7 @@ func (m *ManagerImpl) UnprepareResources(pod *v1.Pod) error {
 			// before this checkpoint succeeds, we will simply call into this (idempotent)
 			// function again.
 			claimInfo.deletePodReference(pod.UID)
+			m.cache.add(&claimInfo)
```
Again -- `add()` as a function name seems weird here.

bart0sh replied:
Agree, but having two identical functions - `add` and `update` doesn't make much sense either. We should come up with one name for both.  My proposal is `addOrUpdate`.

klueska commented on pkg/kubelet/cm/dra/manager.go:
```diff
@@ -356,7 +354,7 @@ func (m *ManagerImpl) UnprepareResources(pod *v1.Pod) error {
 			batches[pluginName] = append(batches[pluginName], claim)
 		}
-		claimInfos[claimInfo.ClaimUID] = claimInfo
+		claimInfos[claimInfo.ClaimUID] = &claimInfo
```
Likewise where about not calling `add()` /  `update()` instead.

bart0sh replied:
We don't want to remove these claim infos at this point. we'll do it after successful `NodeUnprepareResources` call (see line 389 below). This is symmetrical to the `NodePrepareResources`. We first collect claim infos to the map and then adding them to the cache after successful `NodePrepareResources`.

klueska reviewed (commented):
First pass

klueska commented on pkg/kubelet/cm/dra/claiminfo.go:
```diff
@@ -179,11 +186,18 @@ func (cache *claimInfoCache) add(res *ClaimInfo) {
-	return cache.claimInfo[claimName+namespace]
+	claimInfo, ok := cache.claimInfo[claimName+namespace]
+	if !ok {
+		return ClaimInfo{}, fmt.Errorf("unable to get resource for namespace: %s, claim: %s", namespace, claimName)
```
This should not contain the namespace or claimname in the error string. It is the responsibility of the caller to include any passed in parameters in its wrapper error string.

```suggestion
		return ClaimInfo{}, fmt.Errorf("claimInfo not found")
```

klueska replied:
Actually, thinking about this a bit more, should this just return a `bool` instead of an error?

bart0sh replied:
yes, that would be much clearer, I agreee.

bart0sh commented:
@klueska Thank you for the review! I updated the code as suggested, PTAL.

bart0sh commented:
/triage accepted

bart0sh commented:
/retest

klueska commented on pkg/kubelet/cm/dra/claiminfo.go:
```diff
@@ -77,14 +68,30 @@ func (info *ClaimInfo) addCDIDevices(pluginName string, cdiDevices []string) err
+}
+
+func (info *ClaimInfo) deepCopy() ClaimInfo {
+	claimInfo := ClaimInfo{
+		ClaimInfoState: info.ClaimInfoState.DeepCopy(),
+		annotations:    maps.Clone(info.annotations),
 	}
-	return lst
+
+	return claimInfo
 }
```
This is missing the prepared field.

klueska commented on pkg/kubelet/cm/dra/claiminfo.go:
```diff
@@ -122,7 +129,7 @@ func newClaimInfoFromResourceClaim(resourceClaim *resourcev1alpha2.ResourceClaim
 	}
 
-	return newClaimInfo(
+	return *newClaimInfo(
```
For consistency, should we update `newClaimInfo()` to return a non-pointer type as well?

bart0sh replied:
done

klueska commented on pkg/kubelet/cm/dra/claiminfo.go:
```diff
@@ -77,14 +68,30 @@ func (info *ClaimInfo) addCDIDevices(pluginName string, cdiDevices []string) err
+	return cdiDevices
+}
+
+func (info *ClaimInfo) deepCopy() ClaimInfo {
```
By convention, deep copy normally returns a pointer to the new object. We should actually probably rely on https://pkg.go.dev/k8s.io/gengo/examples/deepcopy-gen to generate them for us instead of writing all of the ones in this PR on our own.

bart0sh replied:
Great idea, thank you! done.

klueska reviewed (commented):
In general, this looks good. A few comments.

klueska commented on pkg/kubelet/cm/dra/manager_test.go:
```diff
@@ -798,9 +799,9 @@ func TestPrepareResources(t *testing.T) {
-			if claimInfo == nil {
-				t.Fatalf("claimInfo not found in cache for claim %s", *claimName)
+			claimInfo, ok := manager.cache.get(*claimName, test.pod.Namespace)
+			if !ok {
+				t.Fatalf("can't get claimInfo from cache for the claim %s", *claimName)
```
Let's leave the Fatal string unchanged

bart0sh replied:
done

klueska reviewed (commented):
Small comment, otherwise looks good.

bart0sh commented on pkg/kubelet/cm/dra/claiminfo.go:
```diff
@@ -166,24 +163,36 @@ func newClaimInfoCache(stateDir, checkpointName string) (*claimInfoCache, error)
+		return ClaimInfo{}, true
+	}
+
+	return *deepCopy, true
```
Done, thanks.

bart0sh commented:
/priority important-soon

klueska commented on pkg/kubelet/cm/dra/claiminfo.go:
```diff
@@ -166,24 +163,33 @@ func newClaimInfoCache(stateDir, checkpointName string) (*claimInfoCache, error)
+		return ClaimInfo{}, false
+	}
+
+	deepCopy := ClaimInfo{}
```
```suggestion
	var deepCopy ClaimInfo
```

bart0sh replied:
applied, squashed. PTAL.

klueska commented:
Thanks for this change. It cleans things up quite a bit and is much easier to reason about than requiring locking at two different levels.

/lgtm
/approve

klueska commented on pkg/kubelet/cm/dra/claiminfo.go:
```diff
@@ -166,24 +163,33 @@ func newClaimInfoCache(stateDir, checkpointName string) (*claimInfoCache, error)
 }
 
-func (cache *claimInfoCache) add(res *ClaimInfo) {
+func (cache *claimInfoCache) addOrUpdate(res *ClaimInfo) {
 	cache.Lock()
 	defer cache.Unlock()
 
 	cache.claimInfo[res.ClaimName+res.Namespace] = res
 }
```
/hold

I think we need to update this to *actually* do an update if there is already an entry in the cache (rather than simply overwriting it). Otherwise, previously updated podReferences will be overwritten, for example.

bart0sh replied:
@klueska I've added `addPodReference` and `deletePodReference` APIs to the `claimInfoCache` to do in-place updates. There is still one place in the `PrepareResources` code where I assumed that replacing ClaimInfo object is safe. Not sure you'll like it, so it's just a proposal. Hopefully you'll have an idea how to do it better.

bart0sh replied:
@klueska I'm considering making `addCDIDevices` API to the `claimInfoCache` as well as this is a last one in the `ClaimInfo` that updates its object. Do you think it makes sense?

bart0sh replied:
> Please add unit tests for all of this code...

The code is covered by more than 80%. I'll look if some of the new code is not covered and add some tests if this is the case. Thanks for reminding.

bart0sh replied:
In fact the only thing that should be done on top of this is updating CDI devices after successful `PrepareResources` GRPC call instead of replacing the whole claim. I'll do it today.

pohly replied:
> The code is covered by more than 80%. I'll look if some of the new code is not covered and add some tests if this is the case.

What triggered my "add unit tests" comment was that there was a "addOrUpdate" method but apparently no unit test for the "update" usage of that method. This is a perfect example why the code coverage percentage is not a good indicator for how well something is tested: this method would have been covered 100% with a single, simple test, but that would have ignored all edge cases.

bart0sh replied:
@klueska I've updated the code again to make it more clear and straightforward. Now all updates of cached `ClaimInfo` objects are done through `claimInfoCache` APIs, which lock manager cache. PTAL.

pohly replied:
claimInfoCache doesn't seem to have any unit tests? Testing it indirectly may trigger bugs, but isn't the same as writing dedicated unit tests that can be reviewed to ensure that edge cases are exercised.

At the higher level, https://github.com/kubernetes/kubernetes/blob/9ca60a21fdde35ab01d08827b170759b032921ba/test/e2e_node/dra_test.go#L69-L133 should cover various scenarios for "NodeUnprepareResources must be called":
- kubelet restarts while several unrelated claims are prepared for different pods
- kubelet restarts while several pods use the same claim

klueska replied:
I agree with patrick -- we need standalone unit tests agains the operations on the `claimInfoCache` type itself.

bart0sh replied:
Sure, I'll add unit tests for the claiminfo.go.

bart0sh committed:
kubelet: DRA: add stress test

The tests calls PrepareResources and UnprepareResources API in
parallel to help discover race conditions.

klueska commented on pkg/kubelet/cm/dra/manager.go:
```diff
@@ -108,30 +109,21 @@ func (m *ManagerImpl) PrepareResources(pod *v1.Pod) error {
-		if claimInfo.prepared {
-			// Already prepared this claim, no need to prepare it again
-			continue
+		claimInfo, ok := m.cache.get(*claimName, pod.Namespace)
+		if ok {
+			// Claim exists in cache. Reference the pod.
+			m.cache.addPodReference(*claimName, pod.Namespace, pod.UID)
+			if claimInfo.prepared {
+				// Already prepared this claim, no need to prepare it again
+				continue
+			}
 		}
+		// This saved claim will be used to update ClaimInfo cache
+		// after NodePrepareResources GRPC succeeds
+		resourceClaims[resourceClaim.UID] = resourceClaim
 
 		// Loop through all plugins and prepare for calling NodePrepareResources.
-		for _, resourceHandle := range claimInfo.ResourceHandles {
+		for _, resourceHandle := range resourceClaim.Status.Allocation.ResourceHandles {
```
It feels like this could be done more cleanly. The previous code was very clear (though buggy) -- looking at this new code, it's not as obvious at first glance. Need to think about how to improve it.

bart0sh replied:
I tend to disagree here. I think this is much more clear. claimInfo.ResourceHandles was hiding the source of data, but this code shows it explicitly and clearly.

klueska replied:
This comment is about this whole code block (I highlighted lines 111-126), not resourceHandles specifically.

bart0sh replied:
ah, ok. Thanks for the clarification. How about this one that I posted in another discussion?
```
claimInfo, ok := m.cache.get(*claimName, pod.Namespace)
if ok {
        m.cache.addPodReference(*claimName, pod.Namespace, pod.UID)
}

if claimInfo.prepared {
    continue
}

// I intentionally do it here, after the above check
resourceClaims[resourceClaim.UID] = resourceClaim 
```

It's more readable, but hides that it checks claimInfo.prepared for no reason when claimInfo is not in the cache.

bart0sh replied:
Did as suggested. Thanks.

klueska commented on pkg/kubelet/cm/dra/manager.go:
```diff
@@ -175,27 +167,25 @@ func (m *ManagerImpl) PrepareResources(pod *v1.Pod) error {
+			// NOTE: We update cached claimInfo object here.
+			// This can cause issues if the kubelet gets restarted between
+			// this point and syncToCheckpoint. It will result
 			// in not calling NodeUnprepareResources for this claim
 			// because no claimInfo will be synced back to the cache
-			// for it after the restart. We need to resolve this issue
-			// before moving to beta.
-			m.cache.add(claimInfo)
+			// for it after the restart.
+			// This can be solved by calling `m.cache.syncToCheckpoint`` here
+			// or inside `m.cache.update``
+			if err = m.cache.update(claimInfo.ClaimName, pod.Namespace, pluginName, result.GetCDIDevices()); err != nil {
+				return fmt.Errorf("failed to update claimInfo cache: claim info: %+v, error: %w", claimInfo, err)
+			}
```
Likewise here.

klueska commented on pkg/kubelet/cm/dra/manager.go:
```diff
@@ -304,7 +290,7 @@ func (m *ManagerImpl) GetResources(pod *v1.Pod, container *v1.Container) (*Conta
 func (m *ManagerImpl) UnprepareResources(pod *v1.Pod) error {
 	batches := make(map[string][]*drapb.Claim)
-	claimInfos := make(map[types.UID]*ClaimInfo)
+	claimInfoNames := make(map[types.UID]string)
```
```suggestion
	claimNames := make(map[types.UID]string)
```

klueska replied:
Out of curiosity, though -- why not just track the claimInfo since we are getting it out of the cache anyway? That way we could remain symmetrical with `PrepareResources()`?

bart0sh replied:
The idea was to use it only where it's actually needed. It's still not obvious that claimInfo is a deep copy, which can cause confusion. It's the opposite of what we had - claimInfo was a reference to a cached object, but it was not obvious, and people were updating it without thinking about cache consistency.

klueska commented on pkg/kubelet/cm/dra/manager.go:
```diff
@@ -383,13 +372,14 @@ func (m *ManagerImpl) UnprepareResources(pod *v1.Pod) error {
-			claimInfo := claimInfos[types.UID(claimUID)]
-			claimInfo.deletePodReference(pod.UID)
-			m.cache.delete(claimInfo.ClaimName, pod.Namespace)
+			claimInfoName := claimInfoNames[types.UID(claimUID)]
```
```suggestion
			claimName := claimNames[types.UID(claimUID)]
```

klueska commented on pkg/kubelet/cm/dra/manager.go:
```diff
@@ -419,8 +409,8 @@ func (m *ManagerImpl) PodMightNeedToUnprepareResources(UID types.UID) bool {
 }
 
 // GetCongtainerClaimInfos gets Container's ClaimInfo
```
```suggestion
// GetContainerClaimInfos gets Container's ClaimInfo
```

klueska reviewed (commented):
In general, the changes to move the "update" calls into the cache instead of on the claimInfo type directly, make sense. I have a few comments about the rest of the changes though.

klueska commented on pkg/kubelet/cm/dra/claiminfo.go:
```diff
@@ -30,33 +30,16 @@ import (
-	info.Lock()
-	defer info.Unlock()
-
+func (info *ClaimInfo) updateCDIDevices(pluginName string, cdiDevices []string) error {
```
I would still call this "add" to be symmetrical with `addPodReferences()`
```suggestion
func (info *ClaimInfo) addCDIDevices(pluginName string, cdiDevices []string) error {
```

klueska commented on pkg/kubelet/cm/dra/claiminfo.go:
```diff
@@ -104,15 +95,14 @@ func newClaimInfo(driverName, className string, claimUID types.UID, claimName, n
 
 // newClaimInfoFromResourceClaim creates a new ClaimInfo object
-func newClaimInfoFromResourceClaim(resourceClaim *resourcev1alpha2.ResourceClaim) *ClaimInfo {
+func newClaimInfoFromResourceClaim(resourceClaim *resourcev1alpha2.ResourceClaim, prepared bool, podUIDs sets.Set[string]) ClaimInfo {
```
This API feels awkward

klueska replied:
Let's drop it and add a `prepared` field to the standard `newClaimInfo()` call and just call that everywhere.

bart0sh replied:
Yes, I thought about it. will do, thanks.

bart0sh replied:
done

klueska commented on pkg/kubelet/cm/dra/claiminfo.go:
```diff
@@ -161,36 +153,76 @@ func newClaimInfoCache(stateDir, checkpointName string) (*claimInfoCache, error)
 
-	cache.claimInfo[res.ClaimName+res.Namespace] = res
+	if _, ok := cache.claimInfo[claimInfo.ClaimName+claimInfo.Namespace]; ok {
+		return fmt.Errorf("claim %s, namespace %s: claim info already in cache", claimInfo.ClaimName, claimInfo.Namespace)
```
The caller should (and does) print the details of the info struct passed in:
```suggestion
		return fmt.Errorf("already present")
```

klueska commented on pkg/kubelet/cm/dra/claiminfo.go:
```diff
@@ -161,36 +153,76 @@ func newClaimInfoCache(stateDir, checkpointName string) (*claimInfoCache, error)
+
+	claimInfo, ok := cache.claimInfo[claimName+namespace]
+	if !ok {
+		return fmt.Errorf("claim %s, namespace %s: claim info not in cache", claimName, namespace)
```
```suggestion
		return fmt.Errorf("not present in cache")
```

bart0sh replied:
Why not to do it in one place and force all callers to repeat this? I'd prefer if callers would wrap this error and add their own details to the error.

klueska replied:
The convention for error messages is that the caller adds details of the variables passed in, not the callee.

bart0sh replied:
Can you point me out to the convention? It looks like golang doesn't follow it: https://go.dev/play/p/ARyRBO2gOmJ

klueska replied:
I can't seem to find an appropriate link at the moment. But if you think about it, there must be some convention, otherwise you end up potentially duplicating information at both levels. And the convention I always follow is -- caller includes info about passed in variables, not the callee. The caller knows more about the actual "value" of the variables than the callee, so it is in the position to add better context.

pohly replied:
Go is inconsistent in this regard. Some packages explicitly document that their errors include certain information, like `os` which often includes the path in errors that involved a path.

Without such documentation, the safer assumption is that errors don't include the information and thus caller should wrap errors to add it. This may duplicate information, but at least it's there.

pohly replied:
The guideline that I usually follow is that if the caller cannot add some piece of information, then it has to be in the error generated by the callee. Otherwise I leave it to the caller.

Adding the same piece of information to all errors returned by a function also is an indicator that this should better get added once by wrapping the different errors (less repetitive).

bart0sh replied:
I agree with @klueska that there must be some convention, but I'm not sure this project [follows](https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/apiserver/pkg/admission/config.go#L64) the one that you're proposing here. Generally I somehow dislike forcing callers to repeat adding parameters to the error. In my opinion it's much safer and clearer if callee would do it once and callers would add more info from their context. May be I'm missing something obvious here. Please, enlight me :)

klueska replied:
The error in the callee does not necessarily stem from the use of the parameters passed in, but the caller that's wrapping that error, definitely made the call into the function with those parameters as intended inputs.

For example:
```
var users map[string]int 

func (users Users) updateUserID(username string, id int) error {
    if id < 0 {
        return fmt.Errorf("id less than zero")
    }
    if !match(username, <some pattern>) {
        return fmt.Errorf("username does not match pattern")
    }
    users[username] = id
    return nil
} 

func main() {
  username := "klueska"
  userid := 25
  if err := users.updateUserID(username userid); err != nil {
      fmt.Printf("error updating %s's userID to %d", username, userid)
      os.Exit(1)
  }
}
```

The inner function only has to return the context it knows / cares about at the time it is returning the error and the outer function makes sure to include the context it knows / cares about at the time it is returning / printing its error.

bart0sh replied:
Yes, but if your code calls updateUserID many times, you'll have to repeat adding username and id to the error messages on the caller side instead of simply using the error returned by updateUserID (e.g. "error updating klueska's userID to 25") and adding caller's info on top of it.

bart0sh replied:
Both approaches have their pros and cons. We just need to agree on what to use. It would be better to have the approach we use to be consistent with project (Kubernetes) common practices, but I don't think they exist.

bart0sh replied:
I've updated the code as you've suggested. It doesn't make sense to continue arguing further, I believe.

klueska commented on pkg/kubelet/cm/dra/claiminfo.go:
```diff
@@ -161,36 +153,76 @@ func newClaimInfoCache(stateDir, checkpointName string) (*claimInfoCache, error)
+	claimInfo.prepared = true
+
+	if err := claimInfo.updateCDIDevices(pluginName, cdiDevices); err != nil {
+		return fmt.Errorf("failed to add CDIDevices to claimInfo %+v: %w", claimInfo, err)
```
The error should likely include the list of cdi devices and plugin name

klueska commented on pkg/kubelet/cm/dra/claiminfo.go:
```diff
@@ -161,36 +153,76 @@ func newClaimInfoCache(stateDir, checkpointName string) (*claimInfoCache, error)
+		return fmt.Errorf("claim %s, namespace %s: claim info not in cache", claimName, namespace)
+	}
+
+	claimInfo.prepared = true
```
It feels awkward to be setting this in here

bart0sh replied:
Having more calls would mean keeping cache unlocked between the calls. I tried to avoid doing that. BTW, this is why I called this method `update`. It not only updates CDI devices, but also sets `prepared` field and will possibly do other updates in future.

klueska replied:
I feel pretty strongly that this shouldn't be in here -- especially since we end up calling this once for each pluginName, essentially overwriting this to `true` each time.

bart0sh replied:
If it's not here then this UnprepareResources code
```
			if err = m.cache.update(claimInfo.ClaimName, pod.Namespace, pluginName, result.GetCDIDevices()); err != nil {
				return fmt.Errorf("failed to update claimInfo cache: claim info: %+v, error: %w", claimInfo, err)
			}
```
would look like this:
```
			if err = m.cache.addCDIDevices(claimInfo.ClaimName, pod.Namespace, pluginName, result.GetCDIDevices()); err != nil {
				return fmt.Errorf("failed to update claimInfo cache: claim info: %+v, error: %w", claimInfo, err)
			}
                         // cache is unlocked here and this claim info can be changed from another goroutine
                         if err = m.cache.setPrepared(claimInfo.ClaimName, pod.Namespace, true); err != nil {
                         }
```
So, it lefts cache unlocked in the middle of the update. I intentionally made update as one function to avoid that.

bart0sh replied:
If it should be considered as prepared only after the loop, we should probably change GRPC API to accept slice of claims and call it only once.

klueska replied:
We can't each call goes to a different plugin.

bart0sh replied:
That's a good point. We should update the cache after all calls succeed, but this actually means that we should update cache inside the loop either.

bart0sh replied:
I mean even after the loop it would still be better to update the cache with one lock(`m.cache.update`) than with two (`m.cache.addCDIDevices` and `m.cache.setPrepared`), no?

bart0sh replied:
> We can't -- each call goes to a different plugin.

Yes, but resource shouldn't be considered Prepared if one of the plugins fails to prepare it.

klueska commented on pkg/kubelet/cm/dra/manager.go:
```diff
@@ -67,7 +68,7 @@ func NewManagerImpl(kubeClient clientset.Interface, stateFileDirectory string, n
 // containerResources on success.
 func (m *ManagerImpl) PrepareResources(pod *v1.Pod) error {
 	batches := make(map[string][]*drapb.Claim)
-	claimInfos := make(map[types.UID]*ClaimInfo)
+	resourceClaims := make(map[types.UID]*resourceapi.ResourceClaim)
```
Why not just continue to track claimInfos? That way we can remain symmetrical with `UnprepareResources()`

bart0sh replied:
`UnprepareResources` tracks claim names. Both changes decrease usage of claimInfo objects outside of cache operations. I thought that it's much more clear if claimInfo only used when cache is locked.

klueska replied:
Then I'd argue we should drop `get()` altogether and perform all operations through the cache once a claimInfo has been added (including things like `isPrepared()`.

bart0sh replied:
Can you elaborate? Are you talking about `claimInfoCache.get`? I'd love to drop it, but how is it related to this code? This code was added here to avoid using not cached claim infos.

klueska replied:
Because it feels odd to me to call `get()` to pull something called `claimInfo` out of the cache, and then not use that to track the info about the claim across the calls.

bart0sh replied:
It looks like it's a matter of taste. claimInfo object is used to check `prepared` field. This saves one lock of cache comparing to your code and I did it to decrease a number of cache locks. However, now I understand what you mean. Will try to do what you've suggested.

klueska commented on pkg/kubelet/cm/dra/claiminfo.go:
```diff
@@ -210,6 +242,26 @@ func (cache *claimInfoCache) hasPodReference(UID types.UID) bool {
 	return false
 }
 
+func (cache *claimInfoCache) addPodReference(claimName, namespace string, podUID types.UID) {
+	cache.Lock()
+	defer cache.Unlock()
+
+	claimInfo, ok := cache.claimInfo[claimName+namespace]
+	if ok {
+		claimInfo.PodUIDs.Insert(string(podUID))
+	}
+}
+
+func (cache *claimInfoCache) deletePodReference(claimName, namespace string, podUID types.UID) {
+	cache.Lock()
+	defer cache.Unlock()
+
+	claimInfo, ok := cache.claimInfo[claimName+namespace]
+	if ok {
+		claimInfo.PodUIDs.Delete(string(podUID))
+	}
+}
+
```
Should these return errors if not in the cache?

klueska replied:
To be clear -- I'm not saying they have to, it just feels inconsistent for some of these calls to return an error and others to silently do nothing.

bart0sh replied:
I think it makes sense. Thanks for catching it. With this check the API will look more consistend.

klueska replied:
The alternative is that we don’t lock anywhere inside the claimInfoCahe abstraction itself and instead call m.cache.Lock() / m.cache.Unlock() where appropriate when being used.

klueska replied:
Sorry, this comment but dropped in the wrong place, but you get the point.

bart0sh replied:
> The alternative is that we don’t lock anywhere inside the claimInfoCahe abstraction itself and instead call m.cache.Lock() / m.cache.Unlock() where appropriate when being used.

I thought about it and even tried it in another branch yesterday: https://github.com/bart0sh/kubernetes/commit/8c66f3a9cb7b251fdf9e52809689a93a9868789f

I have mixed fillings about it. It helps to increase number of locks for sure, but it's easier to forget to unlock the cache as unlocking can't be deferred.

klueska replied:
I'm not saying we should do this, but you can use this trick to insert defers at arbitrary points in the code:

Instead of:
```
someCode1()
someCode2()

Lock()
if err := whatever() {
   Unlock()
   return err
}
Unlock()
return nil
```
you can do:
```
someCode1()
someCode2()

func() {
  Lock()
  defer Unlock()
  
  if err := whatever() {
   return err
  }
  return nil
}()
```

bart0sh replied:
Thanks for the tip. Yes, I can do that, but readability would safer. Your first example is much more readable in my opinion. Do you think it makes sense to do?

klueska commented on pkg/kubelet/cm/dra/manager.go:
```diff
@@ -108,30 +109,21 @@ func (m *ManagerImpl) PrepareResources(pod *v1.Pod) error {
-		if claimInfo.prepared {
-			// Already prepared this claim, no need to prepare it again
-			continue
+		claimInfo, ok := m.cache.get(*claimName, pod.Namespace)
+		if ok {
+			// Claim exists in cache. Reference the pod.
+			m.cache.addPodReference(*claimName, pod.Namespace, pod.UID)
+			if claimInfo.prepared {
+				// Already prepared this claim, no need to prepare it again
+				continue
+			}
 		}
+		// This saved claim will be used to update ClaimInfo cache
+		// after NodePrepareResources GRPC succeeds
+		resourceClaims[resourceClaim.UID] = resourceClaim
```
If we do what I suggest above about removing `get()` then this turns into:

```suggestion
// Save a reference to the claim so we can add it to the ClaimInfo
// cache after the NodePrepareResources gRPC call succeeds
resourceClaims[resourceClaim.UID] = resourceClaim

// If the cache already contains info about this claim, then add a pod reference to it
if m.cache.Contains(*claimName, pod.Namespace) {
    m.cache.addPodReference(*claimName, pod.Namespace, pod.UID)
}

// If a claim is already prepared, there is no need to prepare it again
if m.cache.IsPrepared(*claimName, pod.Namespace) {
	continue
}
```

bart0sh replied:
It's the same as a bit changed current code:
```
claimInfo, ok := m.cache.get(*claimName, pod.Namespace)
if ok {
        m.cache.addPodReference(*claimName, pod.Namespace, pod.UID)
}

if claimInfo.prepared {
    continue
}

// I intentionally do it here, after the above check
resourceClaims[resourceClaim.UID] = resourceClaim 
```

However, this code is a bit less clear than the current code as it checks ```prepared``` field for no reason if the claimInfo is not in the cache.

BTW, I agree that getting rid of ```cache.get``` makes sense, but for another reason.

klueska commented on pkg/kubelet/cm/dra/manager.go:
```diff
@@ -175,27 +171,44 @@ func (m *ManagerImpl) PrepareResources(pod *v1.Pod) error {
+					resourceClaim.Namespace,
+					sets.New(string(pod.UID)),
+					resourceHandles,
+					true,
```
false

bart0sh replied:
done, thanks.

klueska commented on pkg/kubelet/cm/dra/manager.go:
```diff
@@ -175,27 +171,44 @@ func (m *ManagerImpl) PrepareResources(pod *v1.Pod) error {
+			if err = m.cache.addCDIDevices(resourceClaim.Name, pod.Namespace, pluginName, result.GetCDIDevices()); err != nil {
+				return fmt.Errorf("claim %s: failed to update claimInfo cache: %w", resourceClaim.Name, err)
+			}
+			if err = m.cache.setPrepared(resourceClaim.Name, pod.Namespace); err != nil {
+				return fmt.Errorf("claim %s: failed to set prepared state: %w", resourceClaim.Name, err)
+			}
```
As discussed before, this should be done outside the loop.

bart0sh replied:
This would require to collect claim names in the loop and then call `m.cache.setPrepared` for each of them. Just want to confirm you understand this.

BTW, should it be done outside of the inner loop (over claims) or outside of the outer loop (over batches)?

klueska replied:
I'd need to look more closely, but the point is that we can't mark the claim as prepared until all aspects of it have been prepared. Otherwise if something fails at any point within this loop, that claim will never be processed again (in plugins that haven't had a chance to see it yet).

bart0sh replied:
done outside of outer loop, PTAL

klueska commented on pkg/kubelet/cm/dra/manager.go:
```diff
@@ -318,10 +327,9 @@ func (m *ManagerImpl) UnprepareResources(pod *v1.Pod) error {
-		claimInfo := m.cache.get(*claimName, pod.Namespace)
-
+		claimInfo, ok := m.cache.get(*claimName, pod.Namespace)
 		// Skip calling NodeUnprepareResource if claim info is not cached
-		if claimInfo == nil {
+		if !ok {
 			continue
 		}
```
This can be replaced with a `contains()` call

bart0sh replied:
claimInfo is actively used in the code below, so if we use `m.cache.contains` here, we'd have to call `m.cache.get` again. Is that what you're suggesting?

klueska commented on pkg/kubelet/cm/dra/manager.go:
```diff
@@ -108,30 +109,25 @@ func (m *ManagerImpl) PrepareResources(pod *v1.Pod) error {
-			// claim does not exist in cache, create new claimInfo object
-			// to be processed later.
-			claimInfo = newClaimInfoFromResourceClaim(resourceClaim)
+		if m.cache.contains(*claimName, pod.Namespace) {
+			// Claim exists in cache. Reference the pod.
+			err = m.cache.addPodReference(*claimName, pod.Namespace, pod.UID)
+			if err != nil {
+				return fmt.Errorf("claim %s: failed to reference pod UID %s: %w", *claimName, pod.UID, err)
+			}
 		}
```
There's a race with the deletion of the cache entry here. I guess that's OK and unavoidable? After failing, the next try will *not* have the claim in the cache and proceed to go down the path of preparing it again.

klueska replied:
Either way, I don't like this and still think we can do something simpler and more readable.

bart0sh replied:
Yes, that's also my understanding.

klueska commented on pkg/kubelet/cm/dra/manager.go:
```diff
@@ -175,27 +171,44 @@ func (m *ManagerImpl) PrepareResources(pod *v1.Pod) error {
+				if len(resourceHandles) == 0 {
+					resourceHandles = make([]resourceapi.ResourceHandle, 1)
+				}
+				claimInfo := newClaimInfo(
+					resourceClaim.Status.DriverName,
+					resourceClaim.Spec.ResourceClassName,
+					resourceClaim.UID,
+					resourceClaim.Name,
+					resourceClaim.Namespace,
+					sets.New(string(pod.UID)),
+					resourceHandles,
+					true,
+				)
+				if err = m.cache.add(&claimInfo); err != nil {
+					return fmt.Errorf("failed to cache claimInfo %+v: %w", claimInfo, err)
+				}
```
Looking at this again, it feels weird to me to have the creation of this entry happen all the way at the end like this (not to mention that it is done inside a loop, where it is only created the first time around the loop).

I prefer the flow we had originally which (with the new abstractions) would be something like the following at the top of this function:
```
if !m.cache.contains() {
    m.cache.newEntry(claim)
}

if err := m.cache.addPodReference(); err != nil {
    return ...
}

if m.cache.isPrepared() {
    continue
)
```
Where `newEntry()` creates the cache entry up front using the logic shown here, and then there is nothing to do at this point. What purpose did moving the logic here serve in your original refactoring?

bart0sh replied:
The reason was to postpone caching new claims until `PrepareResources` GRPC succeeds. Unprepared claimInfo doesn't influence any logic, but creates possibility for a data race.

bart0sh replied:
I'd actually do it even later, outside of the loop, just before checkpointing to decrease data race possibility even more. WDYT?

pohly replied:
I'm not following the discussion in detail but the wording "decrease data race possibility even more" worries me. If we are not sure that the code is free of race conditions, then it should be simplified further until it is simple enough to understand and convince a reader that there are no race conditions.

klueska replied:
Agreed. Ed and I discussed offline and I have put together a new set of changes that ed will be pulling into his PR today.

klueska commented on pkg/kubelet/cm/dra/manager.go:
```diff
@@ -175,29 +171,42 @@ func (m *ManagerImpl) PrepareResources(pod *v1.Pod) error {
+				if err = m.cache.add(&claimInfo); err != nil {
+					return fmt.Errorf("failed to cache claimInfo %+v: %w", claimInfo, err)
+				}
 			}
```
Looking again -- this won't work as is. It is racy with the addPodReference() code above.

Consider the two code blocks (which can run concurrently):
```
if cache.contains(...) {
    cache.addPodReference(...)
}
```
and
```
if cache.contains(...) {
    cache.add(...)
}
```

The following sequence is possible:
```
// pod 0
if cache.contains(claim0) // false
if cache.isPrepared(claim0) // false
if cache.contains(claim0) // false

// pod 1
if cache.contains(claim0) // false
if cache.isPrepared(claim0) // false
if cache.contains(claim0) // false

// pod 0
cache.add(claim0Info) // with reference to pod0

// pod 1
cache.add(claim0Info) // with reference to pod1
```
The entry in the cache now only has a reference to pod1 not pod0

klueska commented on pkg/kubelet/cm/dra/manager.go:
```diff
@@ -175,46 +184,53 @@ func (m *ManagerImpl) PrepareResources(pod *v1.Pod) error {
+					return fmt.Errorf("unable to get claim info for claim %s in namespace %s", claim.Name, claim.Namespace)
+				}
+				if err := info.addCDIDevices(pluginName, result.GetCDIDevices()); err != nil {
+					return fmt.Errorf("unable to add CDI devices for plugin %s claim %s in namespace %s: %v", pluginName, claim.Name, claim.Namespace, err)
```
```suggestion
					return fmt.Errorf("unable to add CDI devices for plugin %s claim %s in namespace %s: %w", pluginName, claim.Name, claim.Namespace, err)
```

klueska commented on pkg/kubelet/cm/dra/manager.go:
```diff
@@ -108,48 +108,57 @@ func (m *ManagerImpl) PrepareResources(pod *v1.Pod) error {
-			pluginName := resourceHandle.DriverName
-			if pluginName == "" {
-				pluginName = resourceClaim.Status.DriverName
+			// Checkpoint to ensure the claim is tracked for this pod before preparing it.
+			if err := m.cache.syncToCheckpoint(); err != nil {
+				return fmt.Errorf("failed to checkpoint claimInfo state: %w", err)
 			}
```
There's still a problem here...

The issue is that the pod reference will get checkpointed here even if the prepare call never succeeds. If the pod is then deleted before it *ever* has a successful prepare call, unprepare will never be called on it, and we will never remove a reference to it from the cache / checkpoint. Future pods that reference the claim will still be able to prepare it, but unprepare will be blocked since we have no mechanism to remove this now deleted pod from the cache / checkpoint.

One way to address this would be to:
1. Move the checkpoint to the end (as it was before)
2. Remove the pod reference from the cache if / when this function returns an error

Let me make the change on my end and push it.

klueska replied:
Pushed.

bart0sh committed:
kubelet: DRA: add unit test for ClaimInfo and claimInfoCache

bart0sh commented:
@pohly @klueska What are we waiting for?

pohly commented:
> What are we waiting for?

For you to say that you are done and need another review... :grin:

pohly commented on pkg/kubelet/cm/dra/claiminfo.go:
```diff
@@ -102,6 +102,10 @@ func (info *ClaimInfo) setCDIDevices(pluginName string, cdiDevices []string) err
 
+	if info.annotations == nil {
+		info.annotations = make(map[string][]kubecontainer.Annotation)
+	}
```
Looks like you found a bug when writing the unit tests?

pohly replied:
I thought you had added them for a reason when adding the unit test (it's in that commit). :shrug:

pohly commented on pkg/kubelet/cm/dra/manager.go:
```diff
@@ -108,48 +108,73 @@ func (m *ManagerImpl) PrepareResources(pod *v1.Pod) error {
+		// claimInfo cache in cases where this function returns an error.
+		defer func(claim *resourceapi.ResourceClaim) {
+			if rerr != nil {
+				m.cache.Lock()
```
Consider this sequence of events:
- PrepareResources is called
- locks the cache
- adds the pod
- somethings fails
- unlocks the cache
- *some other code runs, for example taking a snapshot*
- this `defer` runs, locks again and removes the pod

The other code in the middle sees the pod as being prepared. Isn't that wrong?

klueska replied:
I'm not not concerned with the sequence you mention, but there may still be an issue (see below):

Note that the `podUIDs` list does not track the preparation of the claim. It simply tracks which pods the kubelet is aware of as needing a claim to be prepared. And it only does so in order to ensure that all of the pods that reference the claim call unprepare before the kubelet finally unprepares the claim.

The the race I'd be worried about is if the unprepare call were to _not_ unprepare a resource that it was supposed to. 

So this sequence of events:

* PrepareResources is called by pod0
* locks the cache
* adds reference to pod0
* unlocks the cache
*  starts to prepare the resource

* PrepareResources is called by pod1
* locks the cache
* adds reference to pod1 
* unlocks the cache
*  starts to prepare the resource

* Pod 0 successfully prepares the claim
* locks the cache
* marks claim as prepared
* unlocks the cache

* UnprepareResources is called by pod0
* locks the cache
* sees two pod references
* removes the reference to pod 0
* unlocks the cache
* returns

* Pod 1 fails in the PrepareResources call
* the defer is called
* locks the cache
* removes the reference to pod 1
* unlocks the cache

* pod1 gets deleted before PrepareResources is called again
* UnprepareResources is *not* called for pod1 because it never had a successful PrepareResources call

bart0sh replied:
@klueska Would it help to only add pod reference when the claim is prepared?

pohly replied:
Here's another failure scenario:
- pod uses claims from two drivers
- PrepareResources creates claiminfo with the pod
- NodePrepareResources for driver A succeeds
- NodePrepareResources for driver B fails
- PrepareResources returns and removes the pod from the claiminfo, which is also not marked as "prepared"

If kubelet never retries PrepareResources, how does it know that NodeUnprepareResources must be called for driver A and when does it do that?

I suspect that `ClaimInfo.prepared` is too coarse. Can we turn that into:
- "not prepared" = definitely nothing is prepared
- "partially prepared" = zero or more NodePrepareResources calls were issued. This gets set in the cache *before* the first NodePrepareResources calls and checkpointed.
- "prepared" = all NodePrepareResources calls have completed, set and checkpointed at the end of NodePrepareResources

If PrepareResources gets called again for a partially prepared claim, it needs to do all NodePrepareResources calls again. We rely on drivers to detect that nothing needs to be done.

Same for UnprepareResources.

Is it guaranteed that one or the other gets called? In other words, can it be that we have stale entries in the cache that never get touched again?

I'd like to see some comments in the code that explain not just what PrepareResources do (I can see that in the code!) but also how and when they are called.

bart0sh replied:
@pohly 
> If kubelet never retries PrepareResources, how does it know that NodeUnprepareResources must be called for driver A and when does it do that?

To my understanding the `Kubelet` will retry `PrepareResources` if it fails, which is the case here.

pohly replied:
Will it retry if the function fails as describe above, kubelet gets stopped, the pod gets deleted, and kubelet gets restarted?

klueska replied:
No, it will not.

bart0sh replied:
It will probably not survive a node and a cluster deletion as well :)
It looks like we're trying to fix everything by this PR. Let me remind you that its original purpose was to fix these 2 bugs https://github.com/kubernetes/kubernetes/issues/123474 https://github.com/kubernetes/kubernetes/issues/116411.  I still believe that every revision of this PR was good enough to fix and backport. This doesn't mean that we should stop on that. Trying to come up with a perfect solution puts our users at risk of seeing these data races and checkpoint-related issues in 1.30.

bart0sh replied:
> Will it retry if the function fails as describe above, kubelet gets stopped, the pod gets deleted, and kubelet gets restarted?

That can be done on a Kubelet start if we have enough info/data checkpointed. Again, I'm not sure this PR is a right place to do it.

pohly replied:
Fixing 1.30 is a secondary priority. Our main priority has to be to get this code up to beta quality in 1.31 so that (ideally) we don't need to do redesigns for 1.32.

bart0sh replied:
True, but should it all be done in one PR that suppose to fix a couple of bugs?

pohly replied:
The title of this PR is "fix cache integrity". That is the purpose as far as I am concerned, and it currently doesn't achieve that because we have identified cases where the integrity is not ensured.

bart0sh replied:
I'd argue that this PR does fix cache integrity and worth merging into master and backporting. Calling UnprepareResources when the Kubelet is stopped and the pod is deleted is another issue and worth fixing by another PR. I don't fully agree that fixing data races in 1.30 is a second priority, considerting that people will try to use 1.30 for DRA experiments and the fact that we have plenty of time for 1.31 fixes.

bart0sh replied:
One thing that we still should considering doing in my opinion is to add pod references only when claim is prepared, i.e. when `NodePrepareResources` succeeds for all plugins. That can potentially fix the race that @klueska described above.

klueska replied:
Moving the addition of a pod reference until after the prepare call just flips the problem on its head. It introduces the possibility of an Unprepare call going through too early.

> PrepareResources is called by pod0
> NodePrepareResources succeeds for all plugins
> locks the cache
> adds reference to pod1
> checkpoints the cache
> unlocks the cache
> returns

> PrepareResources is called by pod1
> NodePrepareResources succeeds for all plugins
> (continued below ...)

> UnrepareResources is called by pod0
> NodePrepareResources succeeds for all plugins
> locks the cache
> removes reference to pod0
> NodeUnprepareResources succeeds for all plugins
> checkpoints the cache
> unlocks the cache
> returns

> (continued from above ...)
> locks the cache
> adds reference to pod1
> checkpoints the cache
> unlocks the cache
> returns

Now pod 1 thinks the claim is prepared, but it has actually been unprepared.

That said, we may be able to get away with this suggestion but it will require a bit more logic than simply moving the addition of the podReference until after the plugins have completed successfully.

bart0sh replied:
@klueska Locking claim info before calling `Node[Un]PrepareResources` and keeping it locked until it's fully prepared could be an acceptable trade-off. It would allow to work with other claim infos as cache is not always locked, but wouldn't allow calling GRPC for the same claim. WDYT?

bart0sh commented on pkg/kubelet/cm/dra/claiminfo.go:
```diff
@@ -30,33 +30,66 @@ import (
-	defer info.Unlock()
-
-	info.PodUIDs.Insert(string(podUID))
+// claimInfoCache is a cache of processed resource claims keyed by namespace + claim name.
```
I thought this pair is uniq. Can you give some details?
I agree that using UID makes sense, just want to understand how the claim with the same name can be created in the same namespace.

pohly replied:
It's different incarnations under the same name. First "default/my-claim" gets created and prepared by kubelet. Then kubelet is shut down. A "sufficiently determined" user can now force-delete "default/my-claim", get it allocated again, and when kubelet comes back, "default/my-claim" is not the claim instance that was prepared before.

klueska replied:
This is not a simple change unfortunately. We don't always have the claimUID at the point at which we would like to do a lookup. I think we can get what we need, but it will likely take some restructuring of the cache itself. I'd rather do it as a follow up (or as a commit added on top once we have this core change reviewed).

pohly replied:
I'm fine with adding a TODO and a corresponding issue for tracking.

pohly replied:
Should `namespace + name` be changed to `namespace + "/" + name` already now? Who knows when we get around to using the UID instead.

klueska replied:
Yes. Let me do that.

klueska replied:
Updated along with reworking the commits a bit so they can be better reviewed 1-by-1

bart0sh commented on pkg/kubelet/cm/dra/manager.go:
```diff
@@ -108,48 +108,73 @@ func (m *ManagerImpl) PrepareResources(pod *v1.Pod) error {
+			// If there isn't one yet, then add it to the cache.
+			claimInfo, exists := m.cache.get(resourceClaim.Name, resourceClaim.Namespace)
+			if !exists {
+				claimInfo = m.cache.add(newClaimInfoFromClaim(resourceClaim))
 			}
-			claim := &drapb.Claim{
-				Namespace:      resourceClaim.Namespace,
-				Uid:            string(resourceClaim.UID),
-				Name:           resourceClaim.Name,
-				ResourceHandle: resourceHandle.Data,
+
+			// Add a reference to the current pod in the claim info.
+			// We delay checkpointing of this change until this call
+			// returns successfully. It is OK to do this because we
+			// will only return successfully from this call if the
+			// checkpoint has succeeded. That means if the kubelet is
+			// ever restarted before this checkpoint succeeds, the pod
+			// whose resources are being prepared would never have
+			// started, so it's OK (actually correct) to not include it
+			// in the cache.
+			claimInfo.addPodReference(pod.UID)
```
Would it be better if we create new claimInfo after `NodePrepareResources` succeeds (preferably for all drivers)? This would help to reduce :) chances for a data race. Adding pod reference can be also done only if claimInfo is already prepared.

pohly replied:
See https://github.com/kubernetes/kubernetes/pull/124323#discussion_r1578918305. We need to record that NodePrepareResources may have been called, otherwise we can loose track of that and not call the corresponding NodeUnprepareResources.

klueska replied:
I will do another pass on this tomorrow. The `defer()` was added hastily and was an attempt to avoid needing to either:
1. Introduce a reconcile loop to ensure that unprepare gets called on deleted pods that are still in the podReferences list
2. hold the lock across the calls to prepare/unprepare on the plugin

I really don't want to lock across the plugin callouts, so It seems we may need to introduce a reconcile loop afterall (unless I can come up with something better when I get back to this).

klueska committed:
kubelet: DRA: update locking/checkpoint semantics of the claimInfo cache

Signed-off-by: Kevin Klues <kklues@nvidia.com>

klueska committed:
kubelet: DRA: add a reconcile loop to unprepare claims for deleted pods

Signed-off-by: Kevin Klues <kklues@nvidia.com>

klueska committed:
kubelet: DRA: update manager test to adhere to new claiminfo cache APIs

Signed-off-by: Kevin Klues <kklues@nvidia.com>

bart0sh commented:
All e2e_node test cases pass for this PR:
```
kubernetes (PR142-dra-fix-cache-integrity) $ sudo _output/local/go/bin/e2e_node.test --ginkgo.focus='DRA' --ginkgo.v --report-dir=/tmp/e2e_node.test/logs --node-name=devel --kubelet-flags='--cluster-domain=cluster.local --fail-sw
ap-on=false --cgroup-driver=systemd' --prepull-images=false --container-runtime-endpoint=unix:///run/containerd/containerd.sock --v=3 --kubelet-config-file=test/e2e_node/jenkins/default-kubelet-config.yaml --runtime-config=api/all=true --service-feature-
gates=DynamicResourceAllocation=true --feature-gates=DynamicResourceAllocation=true
...
Ran 10 of 579 Specs in 331.568 seconds
SUCCESS! -- 10 Passed | 0 Failed | 0 Pending | 569 Skipped
PASS
```

@klueska You can get new test cases from [this WIP PR](https://github.com/kubernetes/kubernetes/pull/124617).

klueska committed:
kubelet: DRA: change key of claimInfo cache to "namespace/claimname"

Signed-off-by: Kevin Klues <kklues@nvidia.com>

klueska commented:
> All e2e_node test cases pass for this PR:
> 
> ```
> kubernetes (PR142-dra-fix-cache-integrity) $ sudo _output/local/go/bin/e2e_node.test --ginkgo.focus='DRA' --ginkgo.v --report-dir=/tmp/e2e_node.test/logs --node-name=devel --kubelet-flags='--cluster-domain=cluster.local --fail-sw
> ap-on=false --cgroup-driver=systemd' --prepull-images=false --container-runtime-endpoint=unix:///run/containerd/containerd.sock --v=3 --kubelet-config-file=test/e2e_node/jenkins/default-kubelet-config.yaml --runtime-config=api/all=true --service-feature-
> gates=DynamicResourceAllocation=true --feature-gates=DynamicResourceAllocation=true
> ...
> Ran 10 of 579 Specs in 331.568 seconds
> SUCCESS! -- 10 Passed | 0 Failed | 0 Pending | 569 Skipped
> PASS
> ```
> 
> @klueska You can get new test cases from [this WIP PR](https://github.com/kubernetes/kubernetes/pull/124617).

Are you saying that all of your new test cases in this WIP PR are passing with the changes from this PR?

bart0sh commented:
@klueska 
> Are you saying that all of your new test cases in this WIP PR are passing with the changes from this PR?

Yes, that's what I'm saying. I'm going to add more test cases soon as suggested in [this issue](https://github.com/kubernetes/kubernetes/issues/124608).

bart0sh commented:
@pohly @klueska Please, don't merge it yet. New test case fails and I'm going to investigate why. I'll post my findings here.

bart0sh commented on pkg/kubelet/cm/dra/manager.go:
```diff
@@ -383,19 +509,15 @@ func (m *ManagerImpl) UnprepareResources(pod *v1.Pod) error {
-		if err != nil {
-			return fmt.Errorf("failed to checkpoint claimInfo state, err: %+v", err)
+			m.cache.Lock()
+			m.cache.delete(claimName, namespace)
```
@klueska This would remove the claim from the cache when first UnprepareResources succeeds for it. If there are more than one plugins and second plugin fails UnprepareResources call, the call will never be repeated as the claim is not in the cache anymore.

klueska replied:
Yep, good catch. We only want to delete after the last unprepare call just before syncing to the checkpoint.

klueska replied:
Yep, good catch. This uses to just delete the pod reference and way changed during some iteration of the code. It should be changed back to only removing the pod reference, not the whole claimInfo.

klueska commented on pkg/kubelet/cm/dra/manager.go:
```diff
@@ -404,21 +526,29 @@ func (m *ManagerImpl) UnprepareResources(pod *v1.Pod) error {
 		}
 	}
 
-	// Checkpoint to capture all of the previous deletePodReference() calls.
-	err := m.cache.syncToCheckpoint()
+	// Atomically sync the cache back to the checkpoint.
```
In conjunction with the comment here: https://github.com/kubernetes/kubernetes/pull/124323/files#r1589067383

We should delete the claimInfo at this point just before checkpointing.

klueska committed:
kubelet: DRA: remove check to set pluginName to DriverName if not in ResourceHandle

It has always been validated that a ResourceHandle MUST have DriverName set, so
this check is unnecessary.

Signed-off-by: Kevin Klues <kklues@nvidia.com>

bart0sh commented on pkg/kubelet/cm/dra/manager.go:
```diff
@@ -404,21 +505,35 @@ func (m *ManagerImpl) UnprepareResources(pod *v1.Pod) error {
+	// Atomically perform some operations on the claimInfo cache.
+	err := m.cache.withLock(func() error {
+		// Delete all claimInfos from the cache that have just been unprepared.
+		for _, claimName := range claimNamesMap {
```
@klueska Should we loop over `response.Claims` here?

bart0sh commented:
@klueska I've just pushed e2e changes to this branch. All test cases passed locally:
```
sudo _output/local/go/bin/e2e_node.test --ginkgo.focus='DRA' --ginkgo.v --report-dir=/tmp/e2e_node.test/logs --node-name=devel --kubelet-flags='--cluster-domain=cluster.local --fail-swap-on=false --cgroup-driver=systemd' --prepull-images=false --container-runtime-endpoint=unix:///run/containerd/containerd.sock --kubelet-config-file=test/e2e_node/jenkins/default-kubelet-config.yaml --runtime-config=api/all=true --service-feature-gates=DynamicResourceAllocation=true --feature-gates=DynamicResourceAllocation=true
  W0506 13:15:08.834491  463773 test_context.go:538] Unable to find in-cluster config, using default host : https://127.0.0.1:6443
  I0506 13:15:08.834555 463773 test_context.go:561] The --provider flag is not set. Continuing as if --provider=skeleton had been used.
Running Suite: E2eNode Suite - /home/ed/go/src/k8s.io/kubernetes
================================================================
Random Seed: 1714990508 - will randomize all specs

Will run 16 of 585 specs
------------------------------
[SynchronizedBeforeSuite] 
...
[SynchronizedAfterSuite] PASSED [0.214 seconds]
------------------------------
[ReportAfterSuite] Kubernetes e2e JUnit report
/home/ed/go/src/k8s.io/kubernetes/test/e2e/framework/test_context.go:612
[ReportAfterSuite] PASSED [0.020 seconds]
------------------------------

Ran 16 of 585 Specs in 421.418 seconds
SUCCESS! -- 16 Passed | 0 Failed | 0 Pending | 569 Skipped
PASS

```

bart0sh commented on pkg/kubelet/cm/dra/manager.go:
```diff
@@ -175,46 +266,53 @@ func (m *ManagerImpl) PrepareResources(pod *v1.Pod) error {
+			if !exists {
+				return fmt.Errorf("unable to get claim info for claim %s in namespace %s", claim.Name, claim.Namespace)
+			}
+			info.setPrepared()
```
@klueska We mark claiminfo as prepared only if `NodePrepareResources` succeeded for all plugins. I think we incorrectly handle the case of partial preparation - when the `NodePrepareResources` succeeded only for some plugins. If pod is deleted, Kubelet will not call `NodeUnprepareResources` for those plugins as this information is not cached.

klueska replied:
The `prepared` field is only used to avoid erroneous calls to `PrepareResources` -- `UnprepareResources` doesn't look at it at all. If we have ever created a claimInfo object and added a pod reference to it, we _will_ get an `UnprepareResources` call for it (either in sync with the deletion or in the reconcile loop).

bart0sh replied:
It makes sense to me now. Thanks for the explanations.

bart0sh commented:
/test pull-kubernetes-kind-dra
/test pull-kubernetes-node-e2e-containerd
/test pull-kubernetes-node-e2e-containerd-1-7-dra
/test pull-kubernetes-node-e2e-crio-cgrpv1-dra
/test pull-kubernetes-node-e2e-crio-cgrpv2-dra

bart0sh commented:
/test pull-kubernetes-kind-dra
/test pull-kubernetes-node-e2e-containerd-1-7-dra
/test pull-kubernetes-node-e2e-crio-cgrpv1-dra
/test pull-kubernetes-node-e2e-crio-cgrpv2-dra

bart0sh commented:
/test pull-kubernetes-kind-dra
/test pull-kubernetes-node-e2e-containerd-1-7-dra
/test pull-kubernetes-node-e2e-crio-cgrpv1-dra
/test pull-kubernetes-node-e2e-crio-cgrpv2-dra

bart0sh commented:
/test pull-kubernetes-kind-dra
/test pull-kubernetes-node-e2e-containerd-1-7-dra
/test pull-kubernetes-node-e2e-crio-cgrpv1-dra
/test pull-kubernetes-node-e2e-crio-cgrpv2-dra

bart0sh commented:
/test pull-kubernetes-kind-dra
/test pull-kubernetes-node-e2e-containerd-1-7-dra
/test pull-kubernetes-node-e2e-crio-cgrpv1-dra
/test pull-kubernetes-node-e2e-crio-cgrpv2-dra

bart0sh commented:
@klueska @pohly e2e seem to run without flakes. Can you review this PR again, please?

bart0sh commented:
/assign @pohly

bart0sh commented:
I don't see any comments, so assuming everything is good and going to lgtm this PR to get it merged in a couple of days.

bart0sh commented:
/unhold
/lgtm

bart0sh commented:
@pohly @klueska can you lgtm this PR please?

bart0sh commented:
@klueska Moved new e2e_node test cases back to https://github.com/kubernetes/kubernetes/pull/124617 for further review. Please lgtm this PR, thanks.

klueska commented:
Thanks @bart0sh 
/lgtm

bart0sh commented:
/retest

sontek commented:
## PR Reviewer Guide 🔍

Here are some key observations to aid the review process:

<table>
<tr><td>⏱️&nbsp;<strong>Estimated effort to review</strong>: 4 🔵🔵🔵🔵⚪</td></tr>
<tr><td>🧪&nbsp;<strong>PR contains tests</strong></td></tr>
<tr><td>🔒&nbsp;<strong>No security concerns identified</strong></td></tr>
<tr><td>⚡&nbsp;<strong>Recommended focus areas for review</strong><br><br>

<details><summary><a href='https://github.com/kubernetes/kubernetes/pull/124323/files#diff-09590b07136d966e0c0e99bf17af6b957b7c9ec74892ab7e577422216b237a93R117-R119'><strong>Inverted Test Logic</strong></a>

In `TestNewClaimInfoFromClaim` and `TestNewClaimInfoFromState`, the assertion uses `reflect.DeepEqual` but calls `t.Errorf` when the result *equals* the expected value. The condition should be `!reflect.DeepEqual(result, test.expectedResult)` to fail when they differ. As written, the test passes when the result is wrong and fails when it is correct. Additionally, `TestNewClaimInfoFromClaim` has `Namespace: claimName` in the expected result instead of `Namespace: namespace`, which is a copy-paste error.
</summary>

```go
if reflect.DeepEqual(result, test.expectedResult) {
	t.Errorf("Expected %v, but got %v", test.expectedResult, result)
}
```

</details>

<details><summary><a href='https://github.com/kubernetes/kubernetes/pull/124323/files#diff-203d2de4519315883638490fcafedbe368e5d2dee0cdba88c91cef46614f80acR531-R533'><strong>Lock Misuse in PodMightNeedToUnprepareResources</strong></a>

`PodMightNeedToUnprepareResources` acquires a write lock (`cache.Lock`) before calling `cache.hasPodReference`, which is a read-only operation. This unnecessarily blocks all concurrent readers. More importantly, `hasPodReference` itself does not acquire any lock, so callers that invoke it directly (e.g., in `reconcileLoop` via the cache's internal iteration) rely on the caller holding the appropriate lock. Using a write lock here where a read lock suffices degrades concurrency without benefit.
</summary>

```go
m.cache.Lock()
defer m.cache.Unlock()
return m.cache.hasPodReference(UID)
```

</details>

<details><summary><a href='https://github.com/kubernetes/kubernetes/pull/124323/files#diff-203d2de4519315883638490fcafedbe368e5d2dee0cdba88c91cef46614f80acR440-R474'><strong>Reconcile Loop Deletes Pod References Without Checkpoint</strong></a>

In `unprepareResources`, when `len(claimInfo.PodUIDs) > 1`, `deletePodReference` is called inside `withLock` but the cache is not checkpointed at that point. The comment says this is intentional and the checkpoint happens on successful return. However, the final `withLock` block that calls `syncToCheckpoint` only iterates over `claimNamesMap` (claims that were fully unprepared), not over claims where only a pod reference was removed. If the kubelet restarts after the pod reference deletion but before the final checkpoint, the stale pod reference will persist in the checkpoint and the reconcile loop may attempt to unprepare a claim that is still in use by another pod.
</summary>

```go
	if len(claimInfo.PodUIDs) > 1 {
		// We delay checkpointing of this change until
		// UnprepareResources returns successfully. It is OK to do
		// this because we will only return successfully from this call
		// if the checkpoint has succeeded. That means if the kubelet
		// is ever restarted before this checkpoint succeeds, we will
		// simply call into this (idempotent) function again.
		claimInfo.deletePodReference(podUID)
		return nil
	}

	// This claimInfo name will be used to update ClaimInfo cache
	// after NodeUnprepareResources GRPC succeeds
	claimNamesMap[claimInfo.ClaimUID] = claimInfo.ClaimName

	// Loop through all plugins and prepare for calling NodeUnprepareResources.
	for _, resourceHandle := range claimInfo.ResourceHandles {
		claim := &drapb.Claim{
			Namespace:      claimInfo.Namespace,
			Uid:            string(claimInfo.ClaimUID),
			Name:           claimInfo.ClaimName,
			ResourceHandle: resourceHandle.Data,
		}
		if resourceHandle.StructuredData != nil {
			claim.StructuredResourceHandle = []*resourceapi.StructuredResourceHandle{resourceHandle.StructuredData}
		}
		pluginName := resourceHandle.DriverName
		batches[pluginName] = append(batches[pluginName], claim)
	}

	return nil
})
if err != nil {
	return fmt.Errorf("locked cache operation: %w", err)
}
```

</details>

</td></tr>
</table>
